File: pypp_core/test_dir/python/src/subscriptable_types.py
from pypp_bridge_library_test_0.subscriptable_type import (
    PseudoSubscriptableType,
    PseudoSubscriptableType2,
)


def subscriptable_types_fn():
    print("PYPP SUBSCRIPTABLE TYPES RESULTS:")
    a: PseudoSubscriptableType[int] = PseudoSubscriptableType[int](9)
    a.print()
    b: PseudoSubscriptableType[str] = PseudoSubscriptableType[str](10)
    b.print()
    c: PseudoSubscriptableType2[int] = PseudoSubscriptableType2[int]([1, 2])
    c.print()
    d: PseudoSubscriptableType2[str] = PseudoSubscriptableType2[str](["a", "b"])
    d.print()
    # Showing that you don't need the second '[str]' because of the C++ deduction guide
    #  for PseudoSubscriptableType2 that is in the bridge library.
    e: PseudoSubscriptableType2[str] = PseudoSubscriptableType2(["c", "d"])
    e.print()

File: pypp_cli/test_dir/cpp/src/subscriptable_types.cpp
#include "subscriptable_types.h"
#include "py_list.h"
#include "py_str.h"
#include "pypp_bridge_library_test_0/subscriptable_type.h"
#include "pypp_util/print.h"

void subscriptable_types_fn() {
    pypp::print(pypp::PyStr("pypp SUBSCRIPTABLE TYPES RESULTS:"));
    PseudoSubscriptableTypeCpp<int> a = PseudoSubscriptableTypeCpp<int>(9);
    a.print();
    PseudoSubscriptableTypeCpp<pypp::PyStr> b =
        PseudoSubscriptableTypeCpp<pypp::PyStr>(10);
    b.print();
    PseudoSubscriptableType2Cpp<int> c =
        PseudoSubscriptableType2Cpp<int>(pypp::PyList({1, 2}));
    c.print();
    PseudoSubscriptableType2Cpp<pypp::PyStr> d =
        PseudoSubscriptableType2Cpp<pypp::PyStr>(
            pypp::PyList({pypp::PyStr("a"), pypp::PyStr("b")}));
    d.print();
    PseudoSubscriptableType2Cpp<pypp::PyStr> e = PseudoSubscriptableType2Cpp(
        pypp::PyList({pypp::PyStr("c"), pypp::PyStr("d")}));
    e.print();
}
